### ISSUE TITLE
Option to compile using (G)Lua sytax for CompileFile/CompileString/RunString

### ISSUE THREAD

**Issue by CaptainPRICE** [Facepunch/garrysmod-requests#1341]
GLua (Garry's Lua[JIT]) and vanilla Lua(JIT) are different in a couple of things:
- GLua has C-style comments (`//` and `/* */`)
- GLua has C-style operators (`!`, `&&`, `||`, `!=`)

What if I wanted to compile a piece of code using vanilla Lua (syntax)?

Or, even better usecase (if this option were implemented): we would be able to test which GMod addons (actual code files) use GLua syntax (and on which lines exactly).
Which would be pretty **awesome**, as it could also ease development, and we would be able to write an automated script to convert from GLua into Lua syntax.
We have such cases today, [GMod Lua repository](https://github.com/Facepunch/garrysmod), and [Wire repository](https://github.com/wiremod/wire) are just two of them.
Basically, some codes mix both GLua and Lua syntax, which is against of coding-style (conventions) for the particular project. Going through the manual process of rewritting GLua into Lua syntax is painful for complex projects.
If this option were implemented into [`CompileFile`](https://wiki.garrysmod.com/page/Global/CompileFile)/[`CompileString`](https://wiki.garrysmod.com/page/Global/CompileString)/[`RunString`](https://wiki.garrysmod.com/page/Global/RunString), we would be able to write an automated script that would go through that hard-work itself (by pcalling it, and patching source code by examining error message)!

tl;dr:
Please consider adding this opt-in as a simple `boolean` argument to these functions:
- `CompileFile( string path, boolean vanillaLuaSyntax=false )`
- `CompileString( string code, string identifier, boolean handleError=true, boolean vanillaLuaSyntax=false )`
- `RunString( string code, string identifier="RunString", boolean handleError=true, boolean vanillaLuaSyntax=false )`

#### vanillaLuaSyntax
These functions would still use GLua syntax, by default value (`false`), so it would stay backward compatible.
A user can explicitly specify `true` to force vanilla Lua syntax, in which case if the code that is being compiled contains C-style comment/operator, it would produce an error (`unexpected symbol near`).

**Comment by robotboy655** [comment]
This seems extremely unnecessary for an extremely narrow use case.

C style stuff can be very easily replaced using "Replace All in Files" of any text editor. And I haven't seen many uses of goto, if any at all.

**Comment by Kefta** [comment]
GLua and Lua 5.2 both support ``goto`` and it currently works with RunString:
```
> RunString("goto blah print'foo' ::blah:: print'bar'")...
bar
```
If you want Lua 5.1 syntax support, create your own tokeniser that processes code before running into any code handling functions. Adding this internally would require GMod to have a completely separate Lua state or tokeniser object that supports vanilla Lua syntax.

**Comment by Kefta** [comment]
>No. You SHOULD NOT do that, it is not a reliable way!
>Think about strings and comments, you will be replacing contents in a string, and comments as well.

Have you heard of regex?

**Comment by CaptainPRICE** [comment]
> Have you heard of regex?

Yes, in fact that's what I use (most of the time, sometimes).
And how does Regex solve the problem.. Well, it does not.

Go try it.
Here some file: https://github.com/Facepunch/garrysmod/blob/master/garrysmod/lua/includes/extensions/client/panel/dragdrop.lua

**Comment by CaptainPRICE** [comment]
> GLua and Lua 5.2 both support `goto` and it currently works with RunString

Ah, right, LuaJIT actually has `goto`. I have updated OP (first post).

**Comment by thegrb93** [comment]
>What if I wanted to compile a piece of code using vanilla Lua (syntax)?

Why? And just convert it to vanilla lua. Adding a limiter to gmod's compiler isn't the solution.

**Comment by CaptainPRICE** [comment]
@thegrb93 
> Why?

To avoid repeating of myself...
> Going through the manual process of rewritting GLua into Lua syntax is painful for complex projects.

...

**Comment by Kefta** [comment]
You can use look aheads/behinds to achieve [if-statements](https://www.regular-expressions.info/conditional.html) in Regex. It doesn't replace all matches, it has conditions and can be used to recognise strings and comments.

**Comment by CaptainPRICE** [comment]
@Kefta 
> See https://blog.ostermiller.org/find-comment for an example.

Like I said, altho it can recognize a usage inside a string, it is still **not reliable** (think multi-line string)!
Also, have you read the "Caveats Comments Inside Other Elements" at the bottom of that page?

**Comment by thegrb93** [comment]
>Going through the manual process of rewritting GLua into Lua syntax is painful for complex projects.

You're still going to have to rewrite it lmao. This would only show you were the errors are.

**Comment by Kefta** [comment]
Yes, believe it or not it will require some complex regex to achieve. YOU have to write the fool-proof method to implement it for your specific language, but it is still possible in Regex.

**Comment by CaptainPRICE** [comment]
@thegrb93 

> You're still going to have to rewrite it lmao. This would only show you where the errors are.

Yes? An error message would actually pinpoint it exactly, and the automated script would be smart enough to patch it completely reliably.

**Comment by thegrb93** [comment]
The syntax is probably compiled into the tokenizer and not optional so I highly doubt that. You're way over-complicating this.

**Comment by CaptainPRICE** [comment]
@Kefta 
> Yes, believe it or not it will require some complex regex to achieve.

You really like Regex don't you? Here goes another case against using Regex: Don't forget about String escaping. Play with your Regex, anyways:
https://gist.github.com/CaptainPRICE/17681965616597561d0b1ab914bc11f7

Regex will cease to work...
Code looks that complicated, because it is made smaller to account for all sort of stuff people could do in complex projects, into just a few lines.
As I already said, Regex is not reliable, it simply doesn't and can't understand code, only the parser can. You can stop saying regex now, it is the wrong tool for this task.

Seriously. Why are you all against this option? Am I the only one to see this option much valuable.. 😢

P.S. If you ever come up with the Regex solution that would convert the above code from GLua into Lua syntax (without breaking any of the code including comments), please do let me know.
(*probably never.*)

**Comment by meepen** [comment]
Why do you need Gmod to do this instead of an external lexer and parser?
This seems like you are adding unnecessary complexity for something that
should not be the worry of Garry's mod developers.

**Comment by CaptainPRICE** [comment]
> Why do you need Gmod to do this instead of an external lexer and parser?

The answer is in your question. I don't want to rely/use external lexer/parser for GMod Lua.
Let me be more concrete.
Because of what I am doing is my own "IDE" (such as GCompute & gace), I'd like to implement code warnings when user enables "vanilla Lua syntax" setting in their project, and also to provide a way to convert from GLua into Lua syntax using reliable code-patching technique (which I had described in OP).

**Comment by thegrb93** [comment]
All you have to do is search for garry syntax outside of comments/strings, not hard.

**Comment by Kefta** [comment]
Regex does understand whatever pattern you provide it. The only fault it has is that it can't tackle nesting, but that can be easily solved by using Regex thru a language like JS or whatever you fancy inside of a loop to catch all nested parentheses. Stop saying it's "unreliable" - it can easily detect strings and comments reliably.

**Comment by CaptainPRICE** [comment]
@Kefta 

> Regex does understand whatever pattern you provide it.

"Regex"? Wait, once again?!
You're implying that GMod Lua API offers Regex. It does not. It only has (crappy) "Patterns", inherited from Lua itself.
Well, it does, but only on client-side (using DHTML)... Server-side would require a binary module to use Regex, or a manual library implemented by hand which would have terrible performance btw.

> Stop saying it's "unreliable" - it can easily detect strings and comments reliably.

Yes, it is unreliable for this task, *oh* you just said it by yourself:
> "it can't tackle nesting"

^
<br>

> but that can be easily solved by using Regex thru a language like JS or whatever you fancy inside of a loop to catch all nested parentheses

1. It is inside GMod Lua environment, there's no Regex. *bang*
2. But, speaking from outside, the **intended** "Regex solution" has to be doable using find-replace-all using any code-editor (tell me a list of Regexes you used to convert the GLua into Lua syntax, [code snippet](https://gist.github.com/CaptainPRICE/17681965616597561d0b1ab914bc11f7)). And **not by running 3rd-party program**/code which would use Regex and loops, that would practically be equal to making a parser. *boom*

---

@thegrb93 

> All you have to do is search for garry syntax outside of comments/strings, not hard.

Simple enough. *cough*

**Comment by Kefta** [comment]
Where did I say this has to be done in Lua? This is to convert your code to whatever format you want: you can use whatever language you want. Also, why do you keep adding sound effects at the end of your comments? You're not as witty as you think.

**Comment by meepen** [comment]
This is honestly just a dumb issue and request. They aren't going to ship
two different versions of the luajit lexer because it's a dumb idea and
unnecessary. Make your own or use c to do it with the Lua shared library.

**Comment by AwfulRanger** [comment]
> Because of what I am doing is my own "IDE" (such as GCompute & gace), I'd like to implement code warnings when user enables "vanilla Lua syntax" setting in their project, and also to provide a way to convert from GLua into Lua syntax using reliable code-patching technique (which I had described in OP).

Surely your IDE would have syntax highlighting, right? If you're unable to parse something simple like `//` or `!`, how will you highlight the rest?

**Comment by birdbrainswagtrain** [comment]
> Surely your IDE would have syntax highlighting, right? If you're unable to parse something simple like `//` or `!`, how will you highlight the rest?

This. I'm not really sure how you expect to write a functional IDE with any decent features (syntax highlighting, code completion, etc) without doing some parsing yourself.

**Comment by CaptainPRICE** [comment]
@meepen 
> it captures **all** multiline strings/comments AND regular strings/comments

_Oh_ really, does it?

![image](https://user-images.githubusercontent.com/9789070/57964915-e8aa8a00-793c-11e9-9b2b-c93202f81c5e.png)

Nope. Try harder.

> > @AwfulRanger
> > Surely your IDE would have syntax highlighting, right? If you're unable to parse something simple like `//` or `!`, how will you highlight the rest?
> 
> @birdbrainswagtrain
> This. I'm not really sure how you expect to write a functional IDE with any decent features (syntax highlighting, code completion, etc) without doing some parsing yourself.

Sure there is parsing already.
I am just aiming to avoid re-parsing when it comes to executing the same code (using Lua syntax while user selected GLua syntax), due performance matters.
And caching token info doesn't really help, so.. But, if this option were implemented as I proposed, it would eliminate this redundancy parsing, and at the same time it'd also mean editor performance improvement, because the "tokenizer version"-switch would be handled at C level!

**Comment by SwadicalRag** [comment]
@CaptainPRICE I'm sorry to hear that, and I'm sorry for coming off as rude. I only hoped to show that regex wasn't a simple pattern matching tool.
I'm not involved in this issue in any way, so I'll take my leave.

**Comment by meepen** [comment]
This issue is shit and it seems like you are trying to offload any and all work onto Gmod developers for something that only you would ever use

**Comment by meepen** [comment]
Also your comments about performance is laughable

**Comment by Kefta** [comment]
To amend my comment from earlier,
>The only fault it has is that [Regex] can't tackle nesting,
there are many regex engines that support recursion and will allow you to fully match nesting. These engines include PCRE (for PHP and C), Perl's, and Python's. You can also always use a tool like Pyparse to easily match strings and comments with no pattern work. Your request just seems way outside of the scope of GMod's development: Lua was made to be mutable language for specific applications and there's no guarantee for cross-implementation support.
